Mathlib/Order/JordanHolder.lean

```diff
@@ -624,36 +354,37 @@ theorem trans {s₁ s₂ s₃ : CompositionSeries X} (h₁ : Equivalent s₁ s
     fun i => iso_trans (h₁.choose_spec i) (h₂.choose_spec (h₁.choose i))⟩
 #align composition_series.equivalent.trans CompositionSeries.Equivalent.trans
 
-theorem append {s₁ s₂ t₁ t₂ : CompositionSeries X} (hs : s₁.top = s₂.bot) (ht : t₁.top = t₂.bot)
+protected theorem smash {s₁ s₂ t₁ t₂ : CompositionSeries X}
+    (hs : s₁.last = s₂.head) (ht : t₁.last = t₂.head)
     (h₁ : Equivalent s₁ t₁) (h₂ : Equivalent s₂ t₂) :
-    Equivalent (append s₁ s₂ hs) (append t₁ t₂ ht) :=
+    Equivalent (smash s₁ s₂ hs) (smash t₁ t₂ ht) :=
   let e : Fin (s₁.length + s₂.length) ≃ Fin (t₁.length + t₂.length) :=
     calc
       Fin (s₁.length + s₂.length) ≃ Sum (Fin s₁.length) (Fin s₂.length) := finSumFinEquiv.symm
       _ ≃ Sum (Fin t₁.length) (Fin t₂.length) := Equiv.sumCongr h₁.choose h₂.choose
       _ ≃ Fin (t₁.length + t₂.length) := finSumFinEquiv
-
   ⟨e, by
     intro i
-    refine' Fin.addCases _ _ i
+    refine Fin.addCases ?_ ?_ i
     · intro i
-      simpa [e, top, bot] using h₁.choose_spec i
+      simpa [-smash_toFun, e, smash_castAdd, smash_succ_castAdd] using h₁.choose_spec i
     · intro i
-      simpa [e, top, bot] using h₂.choose_spec i⟩
-#align composition_series.equivalent.append CompositionSeries.Equivalent.append
+      simpa [-smash_toFun, e, smash_natAdd, smash_succ_natAdd] using h₂.choose_spec i⟩
+#align composition_series.equivalent.append CompositionSeries.Equivalent.smash
 
-protected theorem snoc {s₁ s₂ : CompositionSeries X} {x₁ x₂ : X} {hsat₁ : IsMaximal s₁.top x₁}
-    {hsat₂ : IsMaximal s₂.top x₂} (hequiv : Equivalent s₁ s₂)
-    (htop : Iso (s₁.top, x₁) (s₂.top, x₂)) : Equivalent (s₁.snoc x₁ hsat₁) (s₂.snoc x₂ hsat₂) :=
+protected theorem snoc {s₁ s₂ : CompositionSeries X} {x₁ x₂ : X} {hsat₁ : IsMaximal s₁.last x₁}
+    {hsat₂ : IsMaximal s₂.last x₂} (hequiv : Equivalent s₁ s₂)
+    (hlast : Iso (s₁.last, x₁) (s₂.last, x₂)) : Equivalent (s₁.snoc x₁ hsat₁) (s₂.snoc x₂ hsat₂) :=
   let e : Fin s₁.length.succ ≃ Fin s₂.length.succ :=
     calc
       Fin (s₁.length + 1) ≃ Option (Fin s₁.length) := finSuccEquivLast
       _ ≃ Option (Fin s₂.length) := Functor.mapEquiv Option hequiv.choose
       _ ≃ Fin (s₂.length + 1) := finSuccEquivLast.symm
-
   ⟨e, fun i => by
-    refine' Fin.lastCases _ _ i
-    · simpa [e, top] using htop
+    refine Fin.lastCases ?_ ?_ i
+    · simpa [e, show (snoc s₁ x₁ hsat₁).toFun (Fin.last (s₁.length + 1)) = _ from last_snoc _ _ _,
+        show (snoc s₂ x₂ hsat₂).toFun (Fin.last (s₂.length + 1)) = _ from last_snoc _ _ _]
+        using hlast
```

Comment:
All proofs have got smaller, apart from the last couple and the next couple, which seem to have got bigger because Lean 4 has randomly introduced a `.toFun` which wasn't there before. Can you figure out why this is happening? (I can't!)

Reply:
I couldn't figure out where the `toFun` comes from. But I managed to shorten the proof by adding other simp lemma anyway